query-grammar/src/query_grammar.rs

```diff
@@ -120,22 +120,35 @@ fn date_time<'a>() -> impl Parser<&'a str, Output = String> {
 
 fn term_val<'a>() -> impl Parser<&'a str, Output = String> {
     let phrase = char('"').with(many1(satisfy(|c| c != '"'))).skip(char('"'));
-    phrase.or(word())
+    negative_number().or(phrase.or(word()))
 }
 
 fn term_query<'a>() -> impl Parser<&'a str, Output = UserInputLiteral> {
-    let term_val_with_field = negative_number().or(term_val());
-    (field_name(), term_val_with_field).map(|(field_name, phrase)| UserInputLiteral {
-        field_name: Some(field_name),
-        phrase,
+    (field_name(), term_val(), matching_distance_val()).map(
+        |(field_name, phrase, matching_distance)| UserInputLiteral {
+            field_name: Some(field_name),
+            phrase,
+            matching_distance,
+        },
+    )
+}
+
+fn matching_distance_val<'a>() -> impl Parser<&'a str, Output = u32> {
+    let matching_distance = (char('~'), many1(digit())).map(|(_, distance): (_, String)| distance);
+    optional(matching_distance).map(|distance| match distance {
+        Some(d) => d.parse::<u32>().unwrap(),
```

Comment:
can we return an error instead of panic for `"a b"~a`

Reply:
I don't think panics in that case because it would not match the `many1(digit())`.

`"a   b"~1000000000000000000000000000000000000000` on the hand will panic.